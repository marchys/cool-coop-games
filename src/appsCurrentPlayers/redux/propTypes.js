import PropTypes from 'prop-types';

export default {
  currentPlayers: PropTypes.shape({
    id: PropTypes.string.isRequired,
    loading: PropTypes.bool.isRequired,
    loaded: PropTypes.bool.isRequired,
    playerCount: PropTypes.number,
    error: PropTypes.object,
  }),
};
