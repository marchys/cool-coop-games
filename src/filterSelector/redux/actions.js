// Define the ACTION type
export const SET_VISIBILITY_FILTER = 'SET_VISIBILITY_FILTER';

// Define the different states of the ACTION
export const VisibilityFilters = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_PENDING: 'PENDING',
  SHOW_PLAYING: 'PLAYING',
  SHOW_FINISHED: 'FINISHED',
};

// Set the ACTION CREATOR
export function setVisibilityFilter(filter) {
  // Return type and states for the ACTION
  return { type: SET_VISIBILITY_FILTER, filter };
}
