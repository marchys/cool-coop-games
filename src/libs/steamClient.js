import Axios from 'axios';
import config from 'config';

import { addErrorCatching } from './interceptors';

const steamClient = Axios.create({
  baseURL: `${config.steamApi}`,
});

export default addErrorCatching(steamClient);
