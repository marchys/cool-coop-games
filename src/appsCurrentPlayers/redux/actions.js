import { createAction } from 'redux-actions';
import steamClient from 'libs/steamClient';

const fetchCurrentPlayersRequest = createAction('FETCH_CURRENT_PLAYERS_REQUEST');
const fetchCurrentPlayersSuccess = createAction('FETCH_CURRENT_PLAYERS_SUCCESS');
const fetchCurrentPlayersFailure = createAction('FETCH_CURRENT_PLAYERS_FAILURE');

// TODO: not call api if already game in store

const getCurrentPlayers = appId => async dispatch => {
  dispatch(fetchCurrentPlayersRequest({ appId }));

  const { data, error } = await steamClient.get(
    '/ISteamUserStats/GetNumberOfCurrentPlayers/v0001/',
    { params: { appId } },
  );

  if (error) {
    dispatch(fetchCurrentPlayersFailure({ appId, error }));
    return;
  }

  dispatch(fetchCurrentPlayersSuccess({ appId, data: data.response }));
};

export const consts = {
  fetchCurrentPlayersRequest,
  fetchCurrentPlayersSuccess,
  fetchCurrentPlayersFailure,
};

export default {
  getCurrentPlayers,
};
