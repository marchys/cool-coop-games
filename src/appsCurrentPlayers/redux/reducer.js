import produce from 'immer';
import { handleActions } from 'redux-actions';

import { consts } from './actions';

const initialState = {
  loading: false,
  error: false,
  loaded: false,
  apps: [],
};

// TODO: add top loading state validating sub-states

/* eslint-disable no-param-reassign */

export default handleActions(
  {
    [consts.fetchCurrentPlayersRequest]: produce((draft, { payload: { appId } }) => {
      draft.loading = true;
      draft.loaded = false;
      const foundApp = draft.apps.find(app => app.id === appId);
      if (foundApp) {
        foundApp.loading = true;
        foundApp.loaded = false;
        return;
      }
      draft.apps.push({
        id: appId,
        loading: true,
        error: null,
        loaded: false,
      });
    }),
    [consts.fetchCurrentPlayersSuccess]: produce((draft, { payload: { appId, data } }) => {
      const currApp = draft.apps.find(app => app.id === appId);

      currApp.data = data;
      currApp.loading = false;
      currApp.error = null;
      currApp.loaded = true;

      draft.loading = draft.apps.some(app => app.loading);
      draft.loaded = draft.apps.every(app => app.loaded);
    }),
    [consts.fetchCurrentPlayersFailure]: produce((draft, { payload: { appId, error } }) => {
      const currApp = draft.apps.find(app => app.id === appId);

      currApp.error = error;
      currApp.loading = false;
      currApp.loaded = true;

      draft.loading = draft.apps.some(app => app.loading);
      draft.loaded = draft.apps.every(app => app.loaded);
    }),
  },
  initialState,
);

/* eslint-enable no-param-reassign */

export const name = 'appsCurrentPlayers';
