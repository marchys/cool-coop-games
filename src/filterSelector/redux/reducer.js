// Import the ACTION type and states
import { SET_VISIBILITY_FILTER, VisibilityFilters } from './actions';

// Set the default state
const { SHOW_ALL } = VisibilityFilters.SHOW_ALL;

// Handle the SET_VISIBILITY_FILTER ACTION
function visibilityFilter(state = SHOW_ALL, action) {
  switch (action.type) {
    case SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
}

export default visibilityFilter;
