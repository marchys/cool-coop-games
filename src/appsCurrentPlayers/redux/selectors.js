import { createSelector } from 'reselect';
import { name } from './reducer';

const appMapper = ({ data: { player_count: playerCount } = {}, ...rest }) => ({
  ...rest,
  ...(playerCount && { playerCount }),
});

const currentPlayersSelector = state => state[name] || {};

export const loadingSelector = createSelector(currentPlayersSelector, ({ loading }) => loading);

export const appsSelector = createSelector(currentPlayersSelector, ({ apps }) =>
  apps.map(appMapper),
);

export const currentPlayersByAppIdSelector = appId =>
  createSelector(appsSelector, apps => apps.find(app => app.id === appId));

export default {
  loading: loadingSelector,
  apps: appsSelector,
  currentPlayersByAppId: currentPlayersByAppIdSelector,
};
