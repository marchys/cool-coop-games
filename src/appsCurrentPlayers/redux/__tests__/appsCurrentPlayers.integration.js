import { applyMiddleware, createStore, compose, combineReducers } from 'redux';
import thunk from 'redux-thunk';

import steamClient from 'libs/steamClient';

import reducer, { name } from '../reducer';
import actions from '../actions';
import selectors from '../selectors';

jest.mock('libs/steamClient');

function setup(initialState = {}) {
  const middlewares = [thunk];

  const enhancers = [applyMiddleware(...middlewares)];
  const store = createStore(
    combineReducers({ [name]: reducer }),
    initialState,
    compose(...enhancers),
  );

  return {
    store,
  };
}

test('should have initial state', async () => {
  const { store } = setup();

  expect(store.getState()).toMatchSnapshot();
});

describe('get CurrentPlayers', () => {
  beforeEach(async () => {
    steamClient.get.mockClear();
  });

  test('should show loading', async () => {
    steamClient.get.mockReturnValueOnce(new Promise(() => {}));
    const { store } = setup();

    store.dispatch(actions.getCurrentPlayers());

    expect(selectors.loading(store.getState())).toBe(true);
  });

  describe('success', () => {
    const currentPlayers = {
      response: {
        player_count: 22553,
        result: 1,
      },
    };

    const appId = '23451345';

    function setupSuccess(response) {
      steamClient.get.mockImplementationOnce(async () => ({
        data: response,
      }));
    }

    test('should store apps in store', async () => {
      setupSuccess(currentPlayers);
      const { store } = setup();

      await store.dispatch(actions.getCurrentPlayers(appId));

      expect(selectors.apps(store.getState())).toEqual([
        {
          playerCount: 22553,
          error: null,
          id: '23451345',
          loaded: true,
          loading: false,
        },
      ]);
    });

    test('should hide loading', async () => {
      setupSuccess(currentPlayers);
      const { store } = setup();

      await store.dispatch(actions.getCurrentPlayers(appId));

      expect(selectors.loading(store.getState())).toBe(false);
    });
  });

  describe('failure', () => {
    const customError = { error: 'custom error' };

    const appId = '23451345';

    function setupFailure(error) {
      steamClient.get.mockImplementationOnce(async () => ({
        error,
      }));
    }

    beforeEach(async () => {
      steamClient.post.mockClear();
    });

    test('should store error in corresponding app', async () => {
      setupFailure(customError);
      const { store } = setup();

      await store.dispatch(actions.getCurrentPlayers(appId));

      expect(selectors.currentPlayersByAppId(appId)(store.getState()).error).toEqual(customError);
    });

    test('should hide loading', async () => {
      setupFailure(customError);
      const { store } = setup();

      await store.dispatch(actions.getCurrentPlayers(appId));

      expect(selectors.loading(store.getState())).toBe(false);
    });
  });
});
