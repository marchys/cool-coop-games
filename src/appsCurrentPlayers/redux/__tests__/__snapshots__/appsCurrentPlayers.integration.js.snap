// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have initial state 1`] = `
Object {
  "appsCurrentPlayers": Object {
    "apps": Array [],
    "error": false,
    "loaded": false,
    "loading": false,
  },
}
`;
